src/citrine/resources/data_concepts.py

```diff
@@ -363,6 +365,8 @@ def register(self, model: ResourceType, dry_run=False):
         # know how to replace the objects with link-by-uids. loads() converts this string into
         # nested gemd objects, and then the final dumps() converts that to a json-ready string
         # in which all of the object references have been replaced with link-by-uids.
+
+        GEMDJson(scope=CITRINE_SCOPE).dumps(model)  # This apparent no-op populates uids
```

Comment:
Structurally, it might make sense to modify gemd.entity.dict_serializable so that dump generates an object with an ID, but there is value in the non-mutating dump method.

Reply:
Non-blocking: I would still prefer that we just call a method specifically for setting IDs before registering. IMO serialization should not have side-effects and we should just make it easy for the user to meet all the requirements for preserving links through serialization with explicit convenience methods.

Reply:
I generally agree, except I don't know if there's actually a clean solution here.  In either case, this requires a change to the external library.